src/handler/mod.rs

```diff
@@ -730,7 +722,14 @@ impl Handler {
         // If the ENR does not match the observed IP addresses, we consider the Session
         // failed.
         enr.node_id() == node_address.node_id
-            && (enr.udp_socket().is_none() || enr.udp_socket() == Some(node_address.socket_addr))
+            && match node_address.socket_addr {
+                SocketAddr::V4(socket_addr) => enr
+                    .udp4_socket()
+                    .map_or(true, |advertized_addr| socket_addr == advertized_addr),
+                SocketAddr::V6(socket_addr) => enr
+                    .udp6_socket()
+                    .map_or(true, |advertized_addr| socket_addr == advertized_addr),
```

Comment:
I didn't get to the service part yet, but I assume in these edge cases we dont add the ipv4 enrs to the routing table. I will check this when i get to it tho

Reply:
about the first comment, what we ensure here is that the socket we contacted matches the advertised socket, comparing ipv4 against ipv4 and ipv6 against ipv6. This part of the code is among what would get messy if we allow ipv4 addresses to be advertised in ipv6 fields in ENRs. So yeah, a peer contacted via an ipv6 socket that advertises an ipv4 socket will be allowed, as long as the ipv6 field in the ENR is empty, or it's contents match the contacted address, and that the response is sent via the same socket we used to send our own msg.

Reply:
about the second comment, since we only allow contacting nodes that pass the `get_contactable_address` "filter", and that we check the response is received via the selected address, I think this should be ok